lib/xgif/xgif.go

```diff
@@ -0,0 +1,133 @@
+// xgif is a helper package to create GIF animations based on PNG images
+// The resulting animations have the following properties:
+// 1. All frames have the same size (max(pngs.width), max(pngs.height))
+// 2. All PNGs are centered in the given frame
+// 3. The frame background is plain white
+// Note that to convert from a PNG to a GIF compatible image (Bitmap), the PNG image must be quantized (colors are aggregated in median buckets)
+// so that it has at most 255 colors.
+// This is required because GIFs support only 256 colors and we must keep 1 slot for the white background.
+package xgif
+
+import (
+	"bytes"
+	"fmt"
+	"image"
+	"image/color"
+	"image/gif"
+	"image/png"
+
+	"github.com/ericpauley/go-quantize/quantize"
+
+	"oss.terrastruct.com/util-go/go2"
+)
+
+const INFINITE_LOOP = 0
+const BG_INDEX uint8 = 255
+
+var BG_COLOR = color.White
+
+func AnimatePNGs(pngs [][]byte, animIntervalMs int) ([]byte, error) {
+	var width, height int
+	pngImgs := make([]image.Image, len(pngs))
+	for i, pngBytes := range pngs {
+		img, err := png.Decode(bytes.NewBuffer(pngBytes))
+		if err != nil {
+			return nil, err
+		}
+		pngImgs[i] = img
+		bounds := img.Bounds()
+		width = go2.Max(width, bounds.Dx())
+		height = go2.Max(height, bounds.Dy())
+	}
+
+	interval := animIntervalMs / 10 // gif animation interval is in 100ths of a second
+	anim := &gif.GIF{
+		LoopCount: INFINITE_LOOP,
+		Config: image.Config{
+			Width:  width,
+			Height: height,
+		},
+	}
+
+	for _, pngImage := range pngImgs {
+		// 1. convert the PNG into a GIF compatible image (Bitmap) by quantizing it to 255 colors
+		buf := bytes.NewBuffer(nil)
+		err := gif.Encode(buf, pngImage, &gif.Options{
+			NumColors: 255, // GIFs can have up to 256 colors, so keep 1 slot for white background
+			Quantizer: quantize.MedianCutQuantizer{},
+		})
+		if err != nil {
+			return nil, err
+		}
+		gifImg, err := gif.Decode(buf)
+		if err != nil {
+			return nil, err
+		}
+		palettedImg, ok := gifImg.(*image.Paletted)
+		if !ok {
+			return nil, fmt.Errorf("decoded git image could not be cast as *image.Paletted")
+		}
+
+		// 2. make GIF frames of the same size, keeping images centered and with a white background
```

Comment:
its probably faster to do as a SVG post processing step before converting everything to PNG

Reply:
Good idea. Though, this isn't the part that makes it slow.
Of course, this could save some seconds for large exports, but in the examples I shared they were < 1s.
The costly part is converting the PNG to Bitmap
Quantization is quite expensive and then, later on, mapping from the PNG color to the quantized version is also expensive. Basically, ~WxHxC where C is the number of colors (255)